frame/bounties/src/lib.rs

```diff
@@ -653,12 +779,628 @@ decl_module! {
+		/// that a subcurator is not doing their due diligence, and
+		/// we should pick a new subcurator. In this case the subcurator
+		/// deposit is slashed.
+		///
+		/// State of subbounty is moved to Added state
+		/// on successful call completion.
+		///
+		/// - `bounty_id`: ID pair Bounty ID.
+		/// - `subbounty_id`: ID pair SubBounty ID to cancel.
+		#[weight = <T as Config>::WeightInfo::unassign_subcurator()]
+		fn unassign_subcurator(
+			origin,
+			#[compact] bounty_id: BountyIndex,
+			#[compact] subbounty_id: BountyIndex,
+		) {
+			let maybe_sender = ensure_signed(origin.clone())
+				.map(Some)
+				.or_else(|_| T::RejectOrigin::ensure_origin(origin).map(|_| None))?;
+
+			// Ensure parent bounty exist, get active status info.
+			let maybe_bounty_active_status = Self::ensure_bounty_exist(
+				bounty_id,
+				maybe_sender.is_none(),
+			)?;
+
+			// Ensure subbounty is in expected state
+			SubBounties::<T>::try_mutate_exists(
+				bounty_id,
+				subbounty_id,
+				|maybe_subbounty| -> DispatchResult {
+
+					let mut subbounty = maybe_subbounty
+						.as_mut()
+						.ok_or(Error::<T>::InvalidIndex)?;
+
+					let slash_curator = |arg_curator: &T::AccountId,
+						curator_deposit: &mut BalanceOf<T>| {
+							let imbalance = T::Currency::slash_reserved(
+								arg_curator,
+								*curator_deposit,
+							).0;
+							T::OnSlash::on_unbalanced(imbalance);
+							*curator_deposit = Zero::zero();
+						};
+
+					match subbounty.status {
+						SubBountyStatus::Added => {
+							// No curator to unassign at this point.
+							return Err(Error::<T>::UnexpectedStatus.into())
+						}
+						SubBountyStatus::SubCuratorProposed { ref subcurator } => {
+							// A subcurator has been proposed, but not accepted yet.
+							// Either `RejectOrigin`, curator or the proposed subcurator
+							// can unassign the subcurator.
+							ensure!(
+								maybe_sender.map_or(
+									true,
+									|sender| sender == *subcurator ||
+										maybe_bounty_active_status.map_or(false, |v| v.0 == sender)
+								),
+								BadOrigin,
+							);
+
+						},
+						SubBountyStatus::Active { ref subcurator } => {
+							// The bounty is active.
+							match maybe_sender {
+								// If the `RejectOrigin` is calling this function,
+								// slash the subcurator deposit.
+								None => {
+									slash_curator(subcurator, &mut subbounty.curator_deposit);
+									// Continue to change bounty status below...
+								},
+								Some(sender) => {
+									if sender == *subcurator {
+										// This is the subcurator,
+										// willingly giving up their role.
+										// Give back their deposit.
+										T::Currency::unreserve(
+											&subcurator,
+											subbounty.curator_deposit,
+										);
+										// Continue to change bounty status below...
+									} else if maybe_bounty_active_status
+										.as_ref()
+										.map(|v| v.0 == sender)
+										.unwrap()
```

Comment:
expect should have a proof

Reply:
Thanks, @thiolliere.

As suggested, modified the unwrapping process. Now changed to `expect()` instead of `map()`.

Initially thought about it, but considering the context [Control flow reaches here only when call is from signed origin & parent bounty is in active state, & no possibility for `maybe_bounty_active_status` getting None]. But now realized in this context `expect()` is more appropriate & safer than `map()`.

Reply:
ah I missread, ensure_bounty_exist return None only if reject_origin is true, ok then.